# **Pandas**

# **What is Pandas ?**
***The pandas is library in python .which is used for data loading and data manipulating and data analysis.***

- **There are two types of data structure mainly supported by pandas**

**1.Series - 1-D array(list,tuple,set,dictionary).**

**2.DataFrame - 2-D table like excel work sheet.**

# **How to install pandas ?**
- use pip install pandas as pd

**import library**

In [35]:
import pandas as pd # pandas use for loading  and manipulating the data

In [36]:
# load the dataset
df=pd.read_csv("C:\\Users\\Abc\\Downloads\\student_dataset.csv") 
df

StudentID    Name  Gender           Course  Age  Marks Grade       City
0         2001    Isha  Female     Data Science   27     82     F  Bangalore
1         2002   Aarav  Female     Data Science   27     42     A  Hyderabad
2         2003  Vivaan    Male  Web Development   24     98     B    Kolkata
3         2004   Sanya  Female  Web Development   23     52     A  Hyderabad
4         2005   Anaya  Female  Web Development   28     51     A  Hyderabad
..         ...     ...     ...              ...  ...    ...   ...        ...
145       2146   Anaya    Male  Cloud Computing   19     44     B  Bangalore
146       2147   Meera  Female          AI & ML   18     57     D  Hyderabad
147       2148   Kabir  Female  Cloud Computing   21     92     F  Bangalore
148       2149   Sanya  Female   Cyber Security   26     60     D      Delhi
149       2150    Diya  Female          AI & ML   24     44     F      Delhi

[150 rows x 8 columns]

**1.head()** : ***To get first 5 rows in dataset.***

In [38]:
df.head() # to get first 5 rows in data set

StudentID    Name  Gender           Course  Age  Marks Grade       City
0       2001    Isha  Female     Data Science   27     82     F  Bangalore
1       2002   Aarav  Female     Data Science   27     42     A  Hyderabad
2       2003  Vivaan    Male  Web Development   24     98     B    Kolkata
3       2004   Sanya  Female  Web Development   23     52     A  Hyderabad
4       2005   Anaya  Female  Web Development   28     51     A  Hyderabad

**2.tail.()** : ***To get last 5 rows in dataset.***

In [40]:
df.tail() # to get last 5 rows

StudentID   Name  Gender           Course  Age  Marks Grade       City
145       2146  Anaya    Male  Cloud Computing   19     44     B  Bangalore
146       2147  Meera  Female          AI & ML   18     57     D  Hyderabad
147       2148  Kabir  Female  Cloud Computing   21     92     F  Bangalore
148       2149  Sanya  Female   Cyber Security   26     60     D      Delhi
149       2150   Diya  Female          AI & ML   24     44     F      Delhi

**3.describe(include='all')** :***To summarize the data both numerical and texual column.***

In [42]:
df.describe(include='all') # to describe the data both categorical and numerical

StudentID   Name Gender        Course         Age       Marks Grade  \
count    150.000000    150    150           150  150.000000  150.000000   150   
unique          NaN     10      2             5         NaN         NaN     5   
top             NaN  Anaya   Male  Data Science         NaN         NaN     A   
freq            NaN     21     77            36         NaN         NaN    36   
mean    2075.500000    NaN    NaN           NaN   24.346667   67.033333   NaN   
std       43.445368    NaN    NaN           NaN    3.701962   17.598206   NaN   
min     2001.000000    NaN    NaN           NaN   18.000000   40.000000   NaN   
25%     2038.250000    NaN    NaN           NaN   21.250000   52.000000   NaN   
50%     2075.500000    NaN    NaN           NaN   24.000000   62.000000   NaN   
75%     2112.750000    NaN    NaN           NaN   27.000000   82.750000   NaN   
max     2150.000000    NaN    NaN           NaN   30.000000  100.000000   NaN   

             City  
count         150  
unique          5  
top     Bangalore  
freq           48  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

**Name**

- count : The total count of name is 150 this show there is no missing value in this column.
- unique : In Name 10 unique values are present.
- top : The Anaya is most repeated name in the column.
- frequency : The Anaya is 21 times are repeated.

**Gender**
- count : The total count of Gender is 150 this show there is some missing values in this column.
- unique : In Gender 2 unique values are present.(Male and Female)
- top : The Male is most repeated Gender in the column.
- frequency : The Male Gender  is 77 times are repeated.

**Course**
- count : The total count of Course is 150 this show there is no missing values in this column.
- unique : In Course 5 unique values are present.
- top : The Data Science is most repeated name in the column.
- frequency : The Data Science is 77 times are repeated.

**City**
- count : The total count of  city is 150 this show there is no missing values in this column.
- unique : In city column 5 unique cities are present.
- top : The Bangalore is most repeated name in the column.
- frequency : The Bangalore is 48 times are repeated.

In [44]:
df.describe() # to get summary of numerical column

StudentID         Age       Marks
count   150.000000  150.000000  150.000000
mean   2075.500000   24.346667   67.033333
std      43.445368    3.701962   17.598206
min    2001.000000   18.000000   40.000000
25%    2038.250000   21.250000   52.000000
50%    2075.500000   24.000000   62.000000
75%    2112.750000   27.000000   82.750000
max    2150.000000   30.000000  100.000000

**Age**
- mean : The average (mean) Age column is 24.34 .
- std : The standard deviation is 3.70 is less than mean i.e the data values are closer to mean, or around the mean.
- min : The smallest value is 18.
- max : The largest value is 30.
- 25 % (Q1): The 25 % value is 21.25 that show the values below the 25 % of the data fall.
- 50 % (median): The middle value of this age column is 24.
- 75 % (Q3) : 75 % of the product are price at or below 27 with only the high 25 % price above this.

**Marks**
- count : The Marks column has 150  entries .i.e there is no missing value in marks colmn
- mean : The average is 67.033 .
- std : The standard deviation is 17.59 which is less than mean i.e the data values are closer to mean, or around the mean.
- min : The minimum marks is 40.
- 25 % (Q1) : The value is 52 that show the values below the 25 % of the data fall.
- 50 % (median): The middle value of marks column is 62.
- 75 % (Q3) : 75 % of marks at or below 82 with only the high 25 % price above this.
- max : The heigest marks is 100.

**5.shape** : ***To give total number of rows and columns.***

In [47]:
df.shape # shows the number of rows and column

(150, 8)

- There are 150 rows and 8 column are present in this dataset.

**6.info()**: ***summarize the dataframe (i.e shows the column_name,non_null and datatypes).***

In [50]:
df.info() # shows the summary of dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   StudentID  150 non-null    int64 
 1   Name       150 non-null    object
 2   Gender     150 non-null    object
 3   Course     150 non-null    object
 4   Age        150 non-null    int64 
 5   Marks      150 non-null    int64 
 6   Grade      150 non-null    object
 7   City       150 non-null    object
dtypes: int64(3), object(5)
memory usage: 9.5+ KB


- In above StudentID, Age, Marks are integer datatypes.
- Name, Gender, course, Grade, city have object datatypes.

**7.nunique()**: ***It return the total number of unique values in column.***

In [53]:
df.nunique() # that shows the unique values in data

StudentID    150
Name          10
Gender         2
Course         5
Age           13
Marks         56
Grade          5
City           5
dtype: int64

- There  are 150 values of total students.
- 10 unique  names are present.
- 2 genders are their and total 5 courses are available, 13 ages  and 56 marks,total 5 grades and 5 cities are present in this data.

**8.isnull().sum()** :***It check the missing values and count them.***

In [56]:
df.isnull().sum() # isnull() to check for missing values and sum() to get the total count of missing values  

StudentID    0
Name         0
Gender       0
Course       0
Age          0
Marks        0
Grade        0
City         0
dtype: int64

- There is no missing values in the dataset.

In [58]:
df['Course'].nunique() # to get unique value in specific column

5

**9.duplicated()** :***To check duplicate value in column.***

In [60]:
df.duplicated().sum() # shows the duplicate in dataset

0

In [61]:
df['City'].duplicated().sum() # to check specific duplicate in column

145

- There is 145 duplicate cities.

In [63]:
df.drop_duplicates(inplace=True) # to drop the duplicate

**10.rename()**:***To change the column name.***

In [65]:
df.rename(columns={'Course':'designation'},inplace=True) # rename/change the column name

**11.columns** :***To return the column names.***

In [67]:
df.columns

Index(['StudentID', 'Name', 'Gender', 'designation', 'Age', 'Marks', 'Grade',
       'City'],
      dtype='object')

In [68]:
df['designation'].unique() # shows the unique values in Designation

array(['Data Science', 'Web Development', 'AI & ML', 'Cyber Security',
       'Cloud Computing'], dtype=object)

In [69]:
df.nunique()

StudentID      150
Name            10
Gender           2
designation      5
Age             13
Marks           56
Grade            5
City             5
dtype: int64

**12.value_counts()**:***To count the unique value in each column.***

In [71]:
df['designation'].value_counts() # shows the value count in each category

designation
Data Science       36
Web Development    35
AI & ML            34
Cyber Security     25
Cloud Computing    20
Name: count, dtype: int64

- There are 36 students from Data Science course.
- 35 students from Web developments, 34 students from AI & ML , 25 students from cyber security and 20 students from cloud computing.

**13.groupby()** :***To create a group. or Aggregate data by category.***

In [74]:
df.groupby("designation")["Marks"].mean() # to create group

designation
AI & ML            67.911765
Cloud Computing    68.450000
Cyber Security     66.240000
Data Science       70.888889
Web Development    61.971429
Name: Marks, dtype: float64

**14.sort_values()** :***Sorting the rows in ascending or descending oreder.***

In [76]:
df.sort_values(by="Marks",ascending=False) # sort the marks by ascending order

StudentID   Name  Gender      designation  Age  Marks Grade       City
17        2018  Aarav  Female     Data Science   27    100     C    Kolkata
36        2037   Diya  Female  Web Development   29    100     C  Bangalore
10        2011  Aarav  Female     Data Science   26     99     B  Hyderabad
12        2013   Isha  Female  Cloud Computing   20     99     D  Bangalore
69        2070   Diya  Female          AI & ML   23     99     F  Hyderabad
..         ...    ...     ...              ...  ...    ...   ...        ...
62        2063  Aarav    Male  Web Development   23     42     A     Mumbai
128       2129  Aarav  Female  Web Development   30     41     F     Mumbai
45        2046  Anaya    Male          AI & ML   30     41     A    Kolkata
130       2131  Meera    Male  Web Development   23     40     C  Bangalore
107       2108  Sanya    Male     Data Science   24     40     B     Mumbai

[150 rows x 8 columns]

**15.drop()**:***To remove row or column.***

In [78]:
df.drop("StudentID",axis=1,inplace=True) # drop() is used to remove the column

In [79]:
df.drop(0,axis=0,inplace=True)  # dropping rows using index position,axis=0 for rows

**16.index**:***Return the row index.***

In [81]:
df.index 

RangeIndex(start=1, stop=150, step=1)

In [82]:
df[df['Marks'] > 90] # filtring the data where marks > 90

Name  Gender      designation  Age  Marks Grade       City
2    Vivaan    Male  Web Development   24     98     B    Kolkata
5     Aarav    Male  Web Development   28     96     D     Mumbai
10    Aarav  Female     Data Science   26     99     B  Hyderabad
12     Isha  Female  Cloud Computing   20     99     D  Bangalore
17    Aarav  Female     Data Science   27    100     C    Kolkata
30    Rohan  Female     Data Science   24     92     D  Bangalore
36     Diya  Female  Web Development   29    100     C  Bangalore
46    Aryan    Male     Data Science   22     95     F  Hyderabad
47   Vivaan    Male  Web Development   25     91     C  Hyderabad
49     Isha    Male  Web Development   21     94     D  Bangalore
59    Anaya    Male   Cyber Security   18     95     F  Hyderabad
69     Diya  Female          AI & ML   23     99     F  Hyderabad
71    Aarav  Female  Cloud Computing   19     92     B    Kolkata
74    Rohan  Female  Cloud Computing   23     96     F  Bangalore
79    Rohan    Male  Web Development   24     94     D  Hyderabad
104    Isha  Female     Data Science   23     97     B  Bangalore
117   Sanya    Male     Data Science   18     91     D  Bangalore
119   Aryan  Female          AI & ML   25     96     A      Delhi
124   Kabir    Male   Cyber Security   26     97     D    Kolkata
131    Isha    Male     Data Science   30     99     B    Kolkata
147   Kabir  Female  Cloud Computing   21     92     F  Bangalore

In [83]:
df[df['City']=='Delhi']

Name  Gender      designation  Age  Marks Grade   City
6     Sanya    Male     Data Science   18     87     D  Delhi
7     Kabir    Male          AI & ML   26     86     B  Delhi
24    Sanya    Male          AI & ML   26     72     B  Delhi
33   Vivaan    Male     Data Science   24     85     B  Delhi
43     Diya  Female   Cyber Security   28     61     D  Delhi
56    Kabir    Male          AI & ML   22     54     C  Delhi
64    Sanya    Male   Cyber Security   20     56     C  Delhi
67    Kabir    Male  Web Development   25     51     F  Delhi
86     Diya    Male  Web Development   23     67     A  Delhi
99    Aarav  Female          AI & ML   19     65     A  Delhi
106  Vivaan  Female          AI & ML   28     68     D  Delhi
119   Aryan  Female          AI & ML   25     96     A  Delhi
134   Aarav    Male   Cyber Security   23     79     A  Delhi
136    Isha    Male     Data Science   21     55     B  Delhi
140   Aryan    Male  Web Development   24     55     B  Delhi
148   Sanya  Female   Cyber Security   26     60     D  Delhi
149    Diya  Female          AI & ML   24     44     F  Delhi

In [84]:
df[df['Marks'] > 90].shape[0]# filtering the column base on conditon and shap[0]  that count the each student

21

**17.loc[]** :***Access by label.***

In [86]:
df.loc[4] # loc is use for label based indexing.returns rows with the index label 4

Name                     Anaya
Gender                  Female
designation    Web Development
Age                         28
Marks                       51
Grade                        A
City                 Hyderabad
Name: 4, dtype: object

In [87]:
df.loc[:,["Name","Marks"]] # this is shows only Selected column  values

Name  Marks
1     Aarav     42
2    Vivaan     98
3     Sanya     52
4     Anaya     51
5     Aarav     96
..      ...    ...
145   Anaya     44
146   Meera     57
147   Kabir     92
148   Sanya     60
149    Diya     44

[149 rows x 2 columns]

**18.iloc[]** :***Access by position.***

In [89]:
df.iloc[3] # returns the row for the specified index position . 

Name                     Anaya
Gender                  Female
designation    Web Development
Age                         28
Marks                       51
Grade                        A
City                 Hyderabad
Name: 4, dtype: object

In [90]:
df.iloc[:, 0:10] # accessing the records for all the  columns with the index 0-10

Name  Gender      designation  Age  Marks Grade       City
1     Aarav  Female     Data Science   27     42     A  Hyderabad
2    Vivaan    Male  Web Development   24     98     B    Kolkata
3     Sanya  Female  Web Development   23     52     A  Hyderabad
4     Anaya  Female  Web Development   28     51     A  Hyderabad
5     Aarav    Male  Web Development   28     96     D     Mumbai
..      ...     ...              ...  ...    ...   ...        ...
145   Anaya    Male  Cloud Computing   19     44     B  Bangalore
146   Meera  Female          AI & ML   18     57     D  Hyderabad
147   Kabir  Female  Cloud Computing   21     92     F  Bangalore
148   Sanya  Female   Cyber Security   26     60     D      Delhi
149    Diya  Female          AI & ML   24     44     F      Delhi

[149 rows x 7 columns]

In [91]:
df['Additional_score']=df['Marks']+ 5
df['Additional_score']    # to add in 5 each marks and create new column

1       47
2      103
3       57
4       56
5      101
      ... 
145     49
146     62
147     97
148     65
149     49
Name: Additional_score, Length: 149, dtype: int64

In [92]:
df.sort_values(by="Additional_score",ascending=False)

Name  Gender      designation  Age  Marks Grade       City  \
36    Diya  Female  Web Development   29    100     C  Bangalore   
17   Aarav  Female     Data Science   27    100     C    Kolkata   
69    Diya  Female          AI & ML   23     99     F  Hyderabad   
10   Aarav  Female     Data Science   26     99     B  Hyderabad   
131   Isha    Male     Data Science   30     99     B    Kolkata   
..     ...     ...              ...  ...    ...   ...        ...   
1    Aarav  Female     Data Science   27     42     A  Hyderabad   
128  Aarav  Female  Web Development   30     41     F     Mumbai   
45   Anaya    Male          AI & ML   30     41     A    Kolkata   
130  Meera    Male  Web Development   23     40     C  Bangalore   
107  Sanya    Male     Data Science   24     40     B     Mumbai   

     Additional_score  
36                105  
17                105  
69                104  
10                104  
131               104  
..                ...  
1                  47  
128                46  
45                 46  
130                45  
107                45  

[149 rows x 8 columns]

**19.rank()** :***Rank value in a column.***

In [94]:
df['Additional_score'].rank(ascending=False) # rank to change position and return value

1      144.5
2        7.0
3      113.0
4      118.5
5       11.0
       ...  
145    138.0
146     95.0
147     18.0
148     82.5
149    138.0
Name: Additional_score, Length: 149, dtype: float64

**20.astype()** :***To Convert the datatypes of column.***

In [96]:
df['Additional_score'].astype(float) # to change data type of column

1       47.0
2      103.0
3       57.0
4       56.0
5      101.0
       ...  
145     49.0
146     62.0
147     97.0
148     65.0
149     49.0
Name: Additional_score, Length: 149, dtype: float64

**21.apply()**:***Used to apply function to each row or column.***

In [98]:
df['Final_Report']=df['City'].apply(lambda x : "Eligible" if x =='Mumbai' else "not eligible")  # to apply function
df['Final_Report']

1      not eligible
2      not eligible
3      not eligible
4      not eligible
5          Eligible
           ...     
145    not eligible
146    not eligible
147    not eligible
148    not eligible
149    not eligible
Name: Final_Report, Length: 149, dtype: object

In [99]:
df['Final_Result']=df['Marks'].apply(lambda x : "pass" if x >= 75 else "fail")  # to apply function
df['Final_Result']

1      fail
2      pass
3      fail
4      fail
5      pass
       ... 
145    fail
146    fail
147    pass
148    fail
149    fail
Name: Final_Result, Length: 149, dtype: object

**22.query()** :***Use sql like quries to filter data.***

In [101]:
df.query("designation =='Data Science' and Marks > 80") # apply query to filter data

Name  Gender   designation  Age  Marks Grade       City  \
6     Sanya    Male  Data Science   18     87     D      Delhi   
10    Aarav  Female  Data Science   26     99     B  Hyderabad   
17    Aarav  Female  Data Science   27    100     C    Kolkata   
30    Rohan  Female  Data Science   24     92     D  Bangalore   
32   Vivaan  Female  Data Science   21     84     B  Bangalore   
33   Vivaan    Male  Data Science   24     85     B      Delhi   
38   Vivaan  Female  Data Science   27     83     A    Kolkata   
40    Sanya  Female  Data Science   26     85     F  Hyderabad   
46    Aryan    Male  Data Science   22     95     F  Hyderabad   
104    Isha  Female  Data Science   23     97     B  Bangalore   
117   Sanya    Male  Data Science   18     91     D  Bangalore   
131    Isha    Male  Data Science   30     99     B    Kolkata   

     Additional_score  Final_Report Final_Result  
6                  92  not eligible         pass  
10                104  not eligible         pass  
17                105  not eligible         pass  
30                 97  not eligible         pass  
32                 89  not eligible         pass  
33                 90  not eligible         pass  
38                 88  not eligible         pass  
40                 90  not eligible         pass  
46                100  not eligible         pass  
104               102  not eligible         pass  
117                96  not eligible         pass  
131               104  not eligible         pass

**23.pivot_table()** :***Summarize data types of column aggregation function(sum,mean,median,mode).***

In [103]:
df.pivot_table(values="Marks",index='designation',columns='Gender',aggfunc='sum') # to use aggregate function in pivot table

Gender           Female  Male
designation                  
AI & ML            1243  1066
Cloud Computing     820   549
Cyber Security      762   894
Data Science       1384  1086
Web Development     791  1378

In [104]:
df.pivot_table(values="Age",index='designation',columns='Gender',aggfunc='mean') # to use aggregate function in pivot table

Gender              Female       Male
designation                          
AI & ML          23.529412  25.000000
Cloud Computing  22.636364  25.333333
Cyber Security   24.500000  23.230769
Data Science     24.052632  23.625000
Web Development  27.307692  24.409091

# **Correlation**

# **What is correlation ?**
- Correlation refers to the relationship between two or more variables.
- It measure how changes in one variable are linked to the changes in another variable.
- It apply only numerical columns.

# **Type of Correlation**

**1.Positive correlation** :
- The value range is (0-1).it indicates two variables are positively correlated.i.e one variable is increase other variable is also increase or decrease.

**2.Negative correlation** :
- value range is(-1 to 0) . it indicates the two variable are negatively correlated. if one variable is increase the other variable is decrease.

**3.Zero / No correlation** :
- value range is 0. it indicates there is no relationship between variables.

In [106]:
correlation_table=df[['Age','Marks']].corr()# to find correlation
print(correlation_table)

            Age     Marks
Age    1.000000 -0.081903
Marks -0.081903  1.000000


- In Age and Marks column there is no correlation or zero coorelation between them.